### ISSUE TITLE
How to use RepeatClassifier to classify large repeat sequences?

### ISSUE THREAD

**Issue by BoredMa** [Dfam-consortium/RepeatModeler#122]
I am trying to classify a large repeat sequences file(600M+), but I find it is very slow even with a small repeat sequences file(20M+). So, I want to know if there is a fast way to classify a large repeat sequences file? Or maybe, Can I split the file into sub-file and run it in parallel?

**Comment by BoredMa** [comment]
I have tried to run multiple sub-files in parallel, then an error accurs, "begb ERV1-10_AMi-I_2p#LTR/ERV1 MuLV_pol#LTR/ERV1 at /home/hukang/repeat_detect_tools/RepeatModeler-2.0.1/RepeatClassifier line 820, <BXTP> line 103523." Do you know how to solve it?

**Comment by jebrosen** [comment]
> I am trying to classify a large repeat sequences file(600M+), but I find it is very slow even with a small repeat sequences file(20M+).

Are 600M+/20M+ a number of sequences or a total number of nucleotides? How long do you consider slow (and did it even finish on the smaller file?)

> I have tried to run multiple sub-files in parallel, then an error accurs, "begb ERV1-10_AMi-I_2p#LTR/ERV1 MuLV_pol#LTR/ERV1 at /home/hukang/repeat_detect_tools/RepeatModeler-2.0.1/RepeatClassifier line 820, line 103523." Do you know how to solve it?

RepeatClassifier uses a temporary file in the current working directory, so parallel runs will have to be started each in their own working directory. But the overall approach of splitting should otherwise work; each batch will make a relabeled file `....classified` which can be recombined into a single file at the end.

The main program RepeatModeler makes a temporary directory for each run which avoids this parallel-runs problem; that might be something we can do in RepeatClassifier as well.

**Comment by BoredMa** [comment]
Thanks for solving my problem.

**Comment by BoredMa** [comment]
Hi,
I successfully divided the files and ran RepeatClassifier in parallel. But when I merged all the` sub_file.classified` into one, I got `merged_file.classified`. I found that there is a big difference between `merged_file.classified` and  `original_file.classified` which is generated without using partition and paralization . I noticed that many sequences that were marked as unknown in `original_file.classified` are marked as specific categories in `merged_file.classified`. as shown below.

> < >R=8872#Unknown (RR=8858.  TRF=0.000 NSEG=0.000)
< GCACATGGGCATGCATGCACACCCATATACCTGCACACACACATACACAT
< ATACGTACTTTACACACACACACACACGGCTTAAAGGAATGCAGAAGACA
< TTTCCCTGTGGGAACGCACAGCAATGGACGATAATGTTTTTCCTCGGTTC
< CAGTGCCTCCAAGAGAATATGATGATATAATCCTGGCCCCTTTCACCTGG
< GCTTTGCAGCTTGGAAGAAAGAAAAGATATTCCAGTTTGTGCAGTGTTGA
< ACTATCCCTAAAGCAAGTAACAGGAGACGTCAAAATGTATGCTTATGTCA
< CCCTTATGAAAATGCCTTTTTGTTAGAAATTATGAACATCACCCATCACT
< AGGTACAAGGAGGTAATTAGGGTAATTGATTGGCGTTAAATGCAGCAGGG
< ACTGCAAACTGGATTTGCAGAGTCTGGTAGCCGTCGTGGAAAGAA
---
> >R=8872#LINE/L2 (RR=8858.  TRF=0.000 NSEG=0.000)
> TTCTTTCCACGACGGCTACCAGACTCTGCAAATCCAGTTTGCAGTCCCTG
> CTGCATTTAACGCCAATCAATTACCCTAATTACCTCCTTGTACCTAGTGA
> TGGGTGATGTTCATAATTTCTAACAAAAAGGCATTTTCATAAGGGTGACA
> TAAGCATACATTTTGACGTCTCCTGTTACTTGCTTTAGGGATAGTTCAAC
> ACTGCACAAACTGGAATATCTTTTCTTTCTTCCAAGCTGCAAAGCCCAGG
> TGAAAGGGGCCAGGATTATATCATCATATTCTCTTGGAGGCACTGGAACC
> GAGGAAAAACATTATCGTCCATTGCTGTGCGTTCCCACAGGGAAATGTCT
> TCTGCATTCCTTTAAGCCGTGTGTGTGTGTGTGTAAAGTACGTATATGTG
> TATGTGTGTGTGCAGGTATATGGGTGTGCATGCATGCCCATGTGC

We notice that the two sequences are reverse complementary. Why are there different results?

**Comment by jebrosen** [comment]
Did any of these happen to be the last sequence in a file? We recently got a report elsewhere that RepeatClassifier seems to ignore the last entry and this might be the same problem.

**Comment by BoredMa** [comment]
> Did any of these happen to be the last sequence in a file? We recently got a report elsewhere that RepeatClassifier seems to ignore the last entry and this might be the same problem.

Yes, the last sequence of each sub-file is always identified as unknown. Besides this, have you found any other problems？

**Comment by jebrosen** [comment]
As it turns out, this bug was already fixed but we have not yet made a new release of RepeatModeler that includes it. In the meantime, you could use the current in-development version of `RepeatClassifier`, available at this link: <https://raw.githubusercontent.com/Dfam-consortium/RepeatModeler/c2bb4db5134c8517c3e549ddd8d059b06e636ab7/RepeatClassifier>

**Comment by BoredMa** [comment]
> > Did any of these happen to be the last sequence in a file? We recently got a report elsewhere that RepeatClassifier seems to ignore the last entry and this might be the same problem.
> 
> Yes, the last sequence of each sub-file is always identified as unknown. Could you solve the problem and update the code? Besides this, have you found any other problems？

In addition to the last sequence problem, I found another problem: `If I change the order of sequences, the annotation results are different.`  For example, I random the order of A sequences file to get B sequences file, and the differences of A and B classification results are as follows:
>total_same_num=5713, total_diff_num=5, sub_class_diff_num=19, sum=5737
all sequence number=5737

5713 sequences of A and B are totally identical, 5 sequences are totally different, and 19 sequences have the same main class but different sub-class.

**Comment by jebrosen** [comment]
@rmhubley Can you think of any reason for classifications to depend on the order of sequences? Results are generally chosen by best score, which seems like it shouldn't change from run to run or depend on the input order.

**Comment by XingyuLiao** [comment]
We have developed a parallel version of RepeatClassifier based on Spark and greedy glgorithm with dynamic upper boundary, which may solve the problem of slow processing of large-scale repetitive sequences.  Please refer to the paper “Accelerating RepeatClassifier Based on Spark and Greedy Algorithm with Dynamic Upper Boundary“  published online on bioRxiv. https://www.biorxiv.org/content/10.1101/2021.06.03.446998v2 .  The tool is is publicly available at https://github.com/BioinformaticsCSU/SRC .
